HIP/hip-646.md

```diff
@@ -0,0 +1,90 @@
+---
+hip: 646
+title: Fungible Token Metadata Field
+author: May Chan (@rocketmay)
+working-group: Ashe Oro (@Ashe-Oro)
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Review
+created: 2023-01-04
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/627
+updated: 2023-01-04
+---
+
+## Abstract
+
+This HIP proposes the addition of the metadata field to Fungible Tokens (FT), taking after the Non-Fungible Token (NFT) metadata field which was added in HIP-17.
+
+## Motivation
+
+Similar to NFT's on the network, the fungible token implementation is largely functional. There is a need for a token creator to specify additional "metadata" about the token. The data and formatting is explored in HIP-405 but there is no field suitable for storing the link to this metadata. 
+
+The most 'logical' field proposed to store this data is the existing 'memo' field, however a dedicated metadata field is more appropriate.
+
+## Rationale
+
+The success of HIP-412 in standardizing the NFT space on Hedera demonstrates the potential of standardized metadata functionality for FT's. Wallets, explorers and other applications can display relevant supplementary information about tokens to users in a decentralized fashion.
+
+## User stories
+
+As a token creator, I want a platform-standardized method to host information about my token in a decentralized fashion, allowing applications to query data such as token icon and website.
+
+As a wallet or explorer, I want to query a mirror node to obtain the metadata for any fungible token, similar to current non-fungible token calls, to simplify calls and implementation.
+
+As a developer, I want to store the metadata uri in a dedicated field, to avoid ambiguity of using an overloaded field such as 'memo'.
+  
+## Specification
+
+## HAPI Changes
+
+```
+   // Retrieves the metadata of a token
+   rpc getTokenInfo (Query) returns (Response);
+```
+
+### TokenMintTransactionBody
+Property `metadata` is updated to include all tokens.
+```
+message TokenMintTransactionBody {
```

Comment:
I think `TokenCreateTransactionBody` + `TokenUpdateTransactionBody` are more natural places to initialize and update token-level metadata.

Reply:
Np, I'd suggest:

1. Add a new field to [`TokenCreateTransactionBody`](https://github.com/hashgraph/hedera-protobufs/blob/main/services/token_create.proto),
```
message TokenCreateTransactionBody {
    ...
     /**
     * The token metadata, limited to a UTF-8 encoding of no more than 100 bytes;
     * will usually be a URI to a complete description (as NFT metadata generally
     * just links to a HIP-412 compliant JSON document)
     */
    string metadata = 23;
}
```

2. Permit updates by adding a new field to [`TokenUpdateTransactionBody`](https://github.com/hashgraph/hedera-protobufs/blob/main/services/token_update.proto),
```
message TokenUpdateTransactionBody {
    ...
    /**
     * If set, the token's new metadata (UTF-8 encoding max 100 bytes)
     */
    google.protobuf.StringValue metadata = 16;
}
```

3. Support getting the metadata directly from consensus nodes as you already mentioned in the [`getTokenInfo` response](https://github.com/hashgraph/hedera-protobufs/blob/main/services/token_get_info.proto),
```
message TokenInfo {
    ...
    /**
     * The token's metadata, if any.
     */
    string metadata = 27;
}
```

Reply:
Added this. Thanks, @tinker-michaelj

Reply:
Np, thanks @mgarbs!